lib/mpl_toolkits/mplot3d/art3d.py

```diff
@@ -918,15 +937,45 @@ def rotate_axes(xs, ys, zs, zdir):
         return xs, ys, zs
 
 
-def _zalpha(colors, zs):
+def _zalpha(colors, zs, _dscl=None):
     """Modify the alphas of the color list according to depth."""
-    # FIXME: This only works well if the points for *zs* are well-spaced
-    #        in all three dimensions. Otherwise, at certain orientations,
-    #        the min and max zs are very close together.
-    #        Should really normalize against the viewing depth.
+    # Return empty if given empty
     if len(colors) == 0 or len(zs) == 0:
         return np.zeros((0, 4))
-    norm = Normalize(min(zs), max(zs))
-    sats = 1 - norm(zs) * 0.7
+
+    if _dscl is None:
+        # This only works well if the points for *zs* are well-spaced
+        # in all three dimensions. Otherwise, at certain orientations,
+        # the min and max zs are very close together.
+        # Should really normalize against the viewing depth.
+
+        # Normalize the z-depths to the range 0 - 1
+        norm = Normalize(min(zs), max(zs))
+
+        # Generate alpha multipliers using the normalized z-depths so that
+        # closer points
+        # are opaque and the furthest points are still visible, but transparent
+        sats = 1 - norm(zs) * 0.7
+
+    else:
+        # Improved normalization using a scale value derived from the XYZ
+        # limits of the plot
+
+        # Solid near, transparent far, solid default
+        sats = np.clip(1 - (zs - min(zs)) / (_dscl+1e-8), 0.3, 1)
```

Comment:
I'm not really following this, however, why 1e-8?  What if all your data contains small values order 1e-10?

Reply:
I see your point. If the value of 1e-8 is too large relative to the data, then the values in sats will not cover the whole range of transparencies (0.3 to 1) and will instead be biased higher (e.g. 0.5 to 1). Do you have any suggestions for an alternative?

Reply:
I'm realizing that I was only testing with well-behaved values and the calculation for sats is not as robust as I thought.

Reply:
Aha! I remembered why I was doing it that way! It's a little misleading because "zs" sounds like a traditional z value, but it's actually the z-axis of the clipped _3D projection_ of the data, which ends up at a very consistent scale.

Example values for "zs" on a scatterplot of 10 points:
```
[-1.01084182 -1.00841226 -1.00599435 -1.003588   -1.00119315 -0.99880969
 -0.99643756 -0.99407667 -0.99172693 -0.98938829]
[-0.87861664 -0.90297352 -0.92871934 -0.95597639 -0.98488176 -1.01558962
 -1.048274   -1.08313208 -1.12038816 -1.1602985 ]
```

This _does_ vary, but I have never seen it come close to 1e-8 in testing. I forgot since I mainly worked on this a few weeks ago.

Reply:
I guess if zs is normalized to be something visible, then 1e-8 is maybe OK?

Reply:
The value is pretty arbitrarily set at the moment. Is there another value that you think would be more conservative?

Reply:
Again, I'm not really familiar with where you are in the code - you are just trying to avoid a singularity, right?  Would `sats[~np.isfinite(sats)] = 0.3` work?

Reply:
Thanks for the suggestions.

Unfortunately, while the method you're describing could handle the resulting NaN values, it would also result in warning messages because a zero-division would occur first:
```
/home/.../miniconda3/envs/matplotlib_debugging/testing.py:7: RuntimeWarning: divide by zero encountered in true_divide
  sats = np.clip(1 - (zs - min(zs)) / (dscl), 0.3, 1)
/home/.../miniconda3/envs/matplotlib_debugging/testing.py:7: RuntimeWarning: invalid value encountered in true_divide
  sats = np.clip(1 - (zs - min(zs)) / (dscl), 0.3, 1)
```

The goal is just to prevent zero-division when `_dscl==0`, which should only happen when the dataset is empty, or it contains exactly one point. Adding a small value in the denominator is one way to address this, but I suppose just checking the dataset length could work too.

Reply:
...or checking `_dcs` is zero?